File: api/cost_structure_run_versions.py
"""Phase 4 cost-structure run versioning, comparison and approval guards."""
from __future__ import annotations

import json
from datetime import datetime, timezone
from decimal import Decimal
from uuid import uuid4

from flask import Blueprint, jsonify, request
from sqlalchemy import Column, DateTime, Integer, MetaData, String, Table, Text, and_, select

metadata = MetaData()

cost_structure_run_versions = Table(
    "cost_structure_run_versions", metadata,
    Column("id", String(100), primary_key=True),
    Column("project_code", String(100), nullable=False, index=True),
    Column("run_id", String(100), nullable=False, unique=True),
    Column("budget_version_id", String(100), nullable=False, index=True),
    Column("budget_status", String(30), nullable=False),
    Column("direct_cost", String(80), nullable=False),
    Column("total_cost", String(80), nullable=False),
    Column("trace_json", Text, nullable=False),
    Column("created_by", String(100), nullable=False),
    Column("created_at", DateTime(timezone=True), nullable=False),
    Column("row_version", Integer, nullable=False, default=1),
)

APPROVED_STATES = {"APPROVED", "FROZEN", "ARCHIVED"}


class CostStructureRunVersionService:
    def __init__(self, engine):
        self.engine = engine
        metadata.create_all(engine)

    def link(self, project_code: str, run_id: str, budget_version_id: str, budget_status: str,
             direct_cost: str, total_cost: str, trace: dict, actor: str) -> dict:
        project_code = project_code.strip(); run_id = run_id.strip(); budget_version_id = budget_version_id.strip()
        status = budget_status.strip().upper()
        if not project_code or not run_id or not budget_version_id:
            raise ValueError("project_code, run_id and budget_version_id are required")
        if status in APPROVED_STATES:
            raise PermissionError("approved or frozen budget version is read-only")
        Decimal(direct_cost); Decimal(total_cost)
        item_id = str(uuid4()); now = datetime.now(timezone.utc)
        with self.engine.begin() as conn:
            if conn.execute(select(cost_structure_run_versions.c.id).where(cost_structure_run_versions.c.run_id == run_id)).first():
                raise RuntimeError("CONFLICT")
            conn.execute(cost_structure_run_versions.insert().values(
                id=item_id, project_code=project_code, run_id=run_id,
                budget_version_id=budget_version_id, budget_status=status,
                direct_cost=str(direct_cost), total_cost=str(total_cost),
                trace_json=json.dumps(trace, ensure_ascii=False, sort_keys=True),
                created_by=actor, created_at=now, row_version=1,
            ))
        return self.get(run_id)

    def get(self, run_id: str) -> dict:
        with self.engine.connect() as conn:
            row = conn.execute(select(cost_structure_run_versions).where(cost_structure_run_versions.c.run_id == run_id)).mappings().first()
        if not row: raise LookupError("cost structure run version not found")
        item = dict(row); item["trace"] = json.loads(item.pop("trace_json")); item["created_at"] = item["created_at"].isoformat()
        item["deep_link"] = f"/app/cost-structure?project={item['project_code']}&run={run_id}"
        return item

    def compare(self, left_run_id: str, right_run_id: str) -> dict:
        left, right = self.get(left_run_id), self.get(right_run_id)
        if left["project_code"] != right["project_code"]:
            raise ValueError("runs must belong to the same project")
        direct_delta = Decimal(right["direct_cost"]) - Decimal(left["direct_cost"])
        total_delta = Decimal(right["total_cost"]) - Decimal(left["total_cost"])
        return {"project_code": left["project_code"], "left": left_run_id, "right": right_run_id,
                "direct_cost_delta": format(direct_delta, "f"), "total_cost_delta": format(total_delta, "f")}


def build_cost_structure_run_version_blueprint(service: CostStructureRunVersionService, resolve_user_id):
    bp = Blueprint("cost_structure_run_versions", __name__, url_prefix="/api/cost-structures")
    @bp.post("/projects/<project_code>/runs/<run_id>/budget-version")
    def link(project_code: str, run_id: str):
        actor = resolve_user_id()
        if actor is None: return jsonify({"code":"UNAUTHORIZED"}), 401
        body = request.get_json(silent=True) or {}
        try:
            return jsonify(service.link(project_code, run_id, str(body.get("budget_version_id", "")),
                str(body.get("budget_status", "DRAFT")), str(body.get("direct_cost", "0")),
                str(body.get("total_cost", "0")), dict(body.get("trace") or {}), str(actor)))
        except PermissionError as exc: return jsonify({"code":"READ_ONLY","detail":str(exc)}), 409
        except ValueError as exc: return jsonify({"code":"INVALID_ARGUMENT","detail":str(exc)}), 400
        except RuntimeError: return jsonify({"code":"CONFLICT","detail":"run already linked"}), 409
    @bp.get("/runs/compare")
    def compare():
        try: return jsonify(service.compare(request.args.get("left", ""), request.args.get("right", "")))
        except LookupError as exc: return jsonify({"code":"NOT_FOUND","detail":str(exc)}), 404
        except ValueError as exc: return jsonify({"code":"INVALID_ARGUMENT","detail":str(exc)}), 400
    return bp

File: pcces-go/internal/storage/sqlite/cost_structure_run_versions.go
package sqlite

import (
	"context"
	"database/sql"
	"encoding/json"
	"fmt"
	"strings"
	"time"

	errx "github.com/liuxb99/pcces-web/pcces-go/internal/platform/errors"
	"github.com/shopspring/decimal"
)

type CostStructureRunVersion struct {
	ID              string         `json:"id"`
	ProjectCode     string         `json:"project_code"`
	RunID           string         `json:"run_id"`
	BudgetVersionID string         `json:"budget_version_id"`
	BudgetStatus    string         `json:"budget_status"`
	DirectCost      string         `json:"direct_cost"`
	TotalCost       string         `json:"total_cost"`
	Trace           map[string]any `json:"trace"`
	CreatedBy       string         `json:"created_by"`
	CreatedAt       string         `json:"created_at"`
	RowVersion      int64          `json:"row_version"`
	DeepLink        string         `json:"deep_link"`
}

type CostStructureRunDiff struct {
	ProjectCode     string `json:"project_code"`
	Left            string `json:"left"`
	Right           string `json:"right"`
	DirectCostDelta string `json:"direct_cost_delta"`
	TotalCostDelta  string `json:"total_cost_delta"`
}

type CostStructureRunVersionRepository struct{ store *Store }

func NewCostStructureRunVersionRepository(store *Store) *CostStructureRunVersionRepository {
	return &CostStructureRunVersionRepository{store: store}
}

func (r *CostStructureRunVersionRepository) Link(ctx context.Context, item CostStructureRunVersion) (CostStructureRunVersion, error) {
	item.ProjectCode = strings.TrimSpace(item.ProjectCode)
	item.RunID = strings.TrimSpace(item.RunID)
	item.BudgetVersionID = strings.TrimSpace(item.BudgetVersionID)
	item.BudgetStatus = strings.ToUpper(strings.TrimSpace(item.BudgetStatus))
	if item.ProjectCode == "" || item.RunID == "" || item.BudgetVersionID == "" {
		return item, errx.New(errx.CodeInvalidArgument, "project_code, run_id and budget_version_id are required", "P4-COST-006")
	}
	if item.BudgetStatus == "APPROVED" || item.BudgetStatus == "FROZEN" || item.BudgetStatus == "ARCHIVED" {
		return item, errx.New(errx.CodeConflict, "approved or frozen budget version is read-only", "P4-COST-006")
	}
	if _, err := decimal.NewFromString(item.DirectCost); err != nil {
		return item, errx.New(errx.CodeInvalidArgument, "invalid direct_cost", "P4-COST-006")
	}
	if _, err := decimal.NewFromString(item.TotalCost); err != nil {
		return item, errx.New(errx.CodeInvalidArgument, "invalid total_cost", "P4-COST-006")
	}
	trace, err := json.Marshal(item.Trace)
	if err != nil {
		return item, err
	}
	if item.ID == "" {
		item.ID = fmt.Sprintf("CSRV-%d", time.Now().UTC().UnixNano())
	}
	item.CreatedAt = time.Now().UTC().Format(time.RFC3339Nano)
	_, err = r.store.db.ExecContext(ctx, `INSERT INTO cost_structure_run_versions(id,project_code,run_id,budget_version_id,budget_status,direct_cost,total_cost,trace_json,created_by,created_at,row_version) VALUES(?,?,?,?,?,?,?,?,?,?,1)`, item.ID, item.ProjectCode, item.RunID, item.BudgetVersionID, item.BudgetStatus, item.DirectCost, item.TotalCost, string(trace), item.CreatedBy, item.CreatedAt)
	if err != nil {
		return item, errx.New(errx.CodeConflict, "run already linked", "P4-COST-006")
	}
	return r.Get(ctx, item.RunID)
}

func (r *CostStructureRunVersionRepository) Get(ctx context.Context, runID string) (CostStructureRunVersion, error) {
	var item CostStructureRunVersion
	var trace string
	err := r.store.db.QueryRowContext(ctx, `SELECT id,project_code,run_id,budget_version_id,budget_status,direct_cost,total_cost,trace_json,created_by,created_at,row_version FROM cost_structure_run_versions WHERE run_id=?`, runID).Scan(&item.ID, &item.ProjectCode, &item.RunID, &item.BudgetVersionID, &item.BudgetStatus, &item.DirectCost, &item.TotalCost, &trace, &item.CreatedBy, &item.CreatedAt, &item.RowVersion)
	if err == sql.ErrNoRows {
		return item, errx.New(errx.CodeNotFound, "cost structure run version not found", "P4-COST-006")
	}
	if err != nil {
		return item, err
	}
	_ = json.Unmarshal([]byte(trace), &item.Trace)
	item.DeepLink = "/app/cost-structure?project=" + item.ProjectCode + "&run=" + item.RunID
	return item, nil
}

func (r *CostStructureRunVersionRepository) Compare(ctx context.Context, leftID, rightID string) (CostStructureRunDiff, error) {
	left, err := r.Get(ctx, leftID)
	if err != nil {
		return CostStructureRunDiff{}, err
	}
	right, err := r.Get(ctx, rightID)
	if err != nil {
		return CostStructureRunDiff{}, err
	}
	if left.ProjectCode != right.ProjectCode {
		return CostStructureRunDiff{}, errx.New(errx.CodeInvalidArgument, "runs must belong to the same project", "P4-COST-006")
	}
	ld, _ := decimal.NewFromString(left.DirectCost)
	rd, _ := decimal.NewFromString(right.DirectCost)
	lt, _ := decimal.NewFromString(left.TotalCost)
	rt, _ := decimal.NewFromString(right.TotalCost)
	return CostStructureRunDiff{ProjectCode: left.ProjectCode, Left: leftID, Right: rightID, DirectCostDelta: rd.Sub(ld).String(), TotalCostDelta: rt.Sub(lt).String()}, nil
}
